File: algorithms/sorting/insertion_sort/insertion_sort.rb
class Array
  def insertion_sort!
    (1...size).each do |i|
      key = self[i]
      ptr = i - 1
      
      while(ptr >= 0 && key < self[ptr])
        self[ptr + 1] = self[ptr]
        ptr -= 1
      end

      self[ptr + 1] = key
    end
  end
end


def interact
  puts "What size array do you want?"
  size = gets.chomp.to_i

  array = []
  puts "Enter the array : "
  size.times { array.push(gets.chomp.to_i) }

  puts "Current array : "
  puts array.inspect

  array.insertion_sort!

  puts "Sorted array : "
  puts array.inspect
end

interact

File: algorithms/sorting/insertion_sort/insertion_sort.cpp
#include <iostream>

using namespace std;

#define REP(i, start, n) for(int i = start; i < n; i++)

void insertion_sort(int *ar, int size)
{
  int key, ptr;
  REP(i, 1, size) {
    key = ar[i];
    ptr = i - 1;
    while(ptr >= 0 && key < ar[ptr]) {
      ar[ptr + 1] = ar[ptr];
      --ptr;
    }
    ar[ptr + 1] = key;
  }
}

void display_array(int *ar, int size)
{
  REP(i, 0, size) cout << ar[i] << "  ";
  cout << endl;
}

int main()
{
  int *ar, size;

  cout << "What size of array do you want? ";
  cin >> size;

  ar = new int[size];
  cout << "Enter the array : ";
  REP(i, 0, size) cin >> ar[i];

  cout << "Current array : " << endl;
  display_array(ar, size);

  insertion_sort(ar, size);

  cout << "Sorted array : " << endl;
  display_array(ar, size);

  return 0;
}